crates/goose/src/tool_inspection.rs

```diff
@@ -0,0 +1,313 @@
+use anyhow::Result;
+use async_trait::async_trait;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use crate::conversation::message::{Message, ToolRequest};
+use crate::providers::base::Provider;
+
+/// Result of inspecting a tool call
+#[derive(Debug, Clone)]
+pub struct InspectionResult {
+    pub tool_request_id: String,
+    pub action: InspectionAction,
+    pub reason: String,
+    pub confidence: f32,
+    pub inspector_name: String,
+    pub finding_id: Option<String>,
+}
+
+/// Action to take based on inspection result
+#[derive(Debug, Clone, PartialEq)]
+pub enum InspectionAction {
+    /// Allow the tool to execute without user intervention
+    Allow,
+    /// Deny the tool execution completely
+    Deny,
+    /// Require user approval before execution (with optional warning message)
+    RequireApproval(Option<String>),
+}
+
+/// Trait for all tool inspectors
+#[async_trait]
+pub trait ToolInspector: Send + Sync {
+    /// Name of this inspector (for logging/debugging)
+    fn name(&self) -> &'static str;
+
+    /// Inspect tool requests and return results
+    async fn inspect(
+        &self,
+        tool_requests: &[ToolRequest],
+        messages: &[Message],
+        provider: Option<Arc<dyn Provider>>,
+    ) -> Result<Vec<InspectionResult>>;
+
+    /// Whether this inspector is enabled
+    fn is_enabled(&self) -> bool {
+        true
+    }
+
+    /// Priority of this inspector (higher = runs first)
+    fn priority(&self) -> u32 {
+        100
+    }
+}
+
+/// Manages all tool inspectors and coordinates their results
+pub struct ToolInspectionManager {
+    inspectors: Vec<Box<dyn ToolInspector>>,
+}
+
+impl ToolInspectionManager {
+    pub fn new() -> Self {
+        Self {
+            inspectors: Vec::new(),
+        }
+    }
+
+    /// Add an inspector to the manager
+    pub fn add_inspector(&mut self, inspector: Box<dyn ToolInspector>) {
+        self.inspectors.push(inspector);
+        // Sort by priority (highest first)
+        self.inspectors
+            .sort_by(|a, b| b.priority().cmp(&a.priority()));
```

Comment:
that's not a great way to keep a sorted list - I think I would also let the caller of add_inspector determine the priority not have it be self declared